src/Parallel/Algorithm.hpp

```diff
@@ -1121,6 +1154,13 @@ bool AlgorithmImpl<ParallelComponent, tmpl::list<PhaseDepActionListsPack...>>::
   }
   return true;
 }
+
+template <typename ParallelComponent, typename... PhaseDepActionListsPack>
+void AlgorithmImpl<ParallelComponent, tmpl::list<PhaseDepActionListsPack...>>::
+    initiate_shutdown(const std::exception& /*exception*/) {
+  CkError("\nException caught in Algorithm:\n\n\nShutting down...\n");
```

Comment:
that would print the long error messages twice...

Reply:
Ah, because the message is already printed in `ERROR` and `ASSERT`? Then perhaps print `e.what()` here and remove the `CkError()` in `ERROR` and `ASSERT`? Otherwise, I think regularly thrown exceptions won't get printed anywhere.

Reply:
I do not want to move printing the error message outside of ABORT/ASSERT in case there is a code path outside the try-catch's that I've added.

Reply:
How about ERROR and ASSERT throw a `spectre_error` or something instead of a generic `runtime_error`, and print the message if it's anything other than that?

Reply:
I agree, we should have `spectre_error` (or `SpectreError`) and `Assert` exceptions. We could have `Assert` inherit from `Error` or something, so that way code that doesn't care about error vs. assert can catch one, while code that does care about it can catch or `Assert`. I think throwing something other than `runtime_error` is really important because in the root finders we care about the type of exception we catch: if Boost throws an exception we catch it and try to recover the root find (possibly), but if it's a SpECTRE error or assert that means things have gone fatally wrong and recovery could be difficult or impossible.

I propose adding `SpectreError` and `SpectreAssert` exception classes and throwing them in the macros.

It's not clear to me that printing inside `ERROR`/`ASSERT` is the right thing to do. Wouldn't it be favorable to print the exception in all cases? If uncaught exceptions are not printed, then I agree that we should print inside the macros, otherwise I think we should only print in the try-catch. If we have to print in the macros then I think Will's suggestion of print any non-spectre exceptions should be made because I will definitely get annoying having something print "Yo dawg, somethin's wrong. Good luck!"